File: stationarity_diagnostics.R
library(urca)

rodar_adf_pfaff <- function(data_ts) {
  # Cria matrizes para armazenar estatísticas e valores críticos
  teststat <- matrix(NA, nrow = ncol(data_ts), ncol = 6)
  colnames(teststat) <- c('tau3', 'phi2', 'phi3', 'tau2', 'phi1', 'tau1')
  rownames(teststat) <- colnames(data_ts)
  
  cval <- matrix(NA, ncol = 3, nrow = 6)
  colnames(cval) <- c('1pct','5pct','10pct')
  rownames(cval) <- c('tau3','phi2','phi3','tau2', 'phi1', 'tau1')
  
  results <- matrix(NA, ncol = 1, nrow = ncol(data_ts))
  colnames(results) <- 'ProcessoGerador'
  rownames(results) <- colnames(data_ts)
  
  for (i in 1:ncol(data_ts)) {
    serie <- data_ts[, i]
    
    # Trend (com tendência e constante)
    adf_trend <- ur.df(serie, type = 'trend', selectlags = 'AIC')
    teststat[i, 1:3] <- adf_trend@teststat[1:3]
    cval[1:3, ] <- adf_trend@cval[1:3, ]
    
    # Drift (com constante)
    adf_drift <- ur.df(serie, type = 'drift', selectlags = 'AIC')
    teststat[i, 4:5] <- adf_drift@teststat[1:2]
    cval[4:5, ] <- adf_drift@cval[1:2, ]
    
    # None (sem constante nem tendência)
    adf_none <- ur.df(serie, type = 'none', selectlags = 'AIC')
    teststat[i, 6] <- adf_none@teststat[1]
    cval[6, ] <- adf_none@cval[1, ]
    
    # Protocolo de Pfaff (2008)
    if (teststat[i, 1] < cval[1,1] | teststat[i, 1] < cval[1,2] | teststat[i, 1] < cval[1,3]) {
      results[i, ] <- 'Estacionário ao Redor de uma Tendência Linear'
    } else if (teststat[i, 3] > cval[3,1] | teststat[i, 3] > cval[3,2] | teststat[i, 3] > cval[3,3]) {
      results[i, ] <- 'Passeio Aleatório com Tendência Linear'
    } else if (teststat[i, 4] < cval[4,1] | teststat[i, 4] < cval[4,2] | teststat[i, 4] < cval[4,3]) {
      results[i, ] <- 'Estacionário ao Redor de uma Constante'
    } else if (teststat[i, 5] > cval[5,1] | teststat[i, 5] > cval[5,2] | teststat[i, 5] > cval[5,3]) {
      results[i, ] <- 'Passeio Aleatório com Drift'
    } else if (teststat[i, 6] < cval[6,1] | teststat[i, 6] < cval[6,2] | teststat[i, 6] < cval[6,3]) {
      results[i, ] <- 'Estacionário com Média Zero'
    } else {
      results[i, ] <- 'Passeio Aleatório sem Drift'
    }
  }
  
  return(list(estatisticas = teststat, criticos = cval, resultado = results))
}

File: stationarity_diagnostics.py

import pandas as pd
import numpy as np
from statsmodels.tsa.stattools import adfuller





def rodar_adf_pfaff(df):
    """
    Roda o protocolo de Pfaff (2008) para detecção de raiz unitária
    em cada coluna de um DataFrame contendo séries temporais.

    Parâmetros:
    df (pd.DataFrame): DataFrame com séries temporais nas colunas.

    Retorna:
    pd.DataFrame: Classificação do processo gerador de cada série.
    """

    resultados = {}

    for col in df.columns:
        serie = df[col].dropna()

        # Etapa 1 - Teste com tendência linear (regression='ct')
        resultado_ct = adfuller(serie, regression='ct', autolag='AIC')
        stat_ct = resultado_ct[0]
        crit_ct = resultado_ct[4]

        if stat_ct < crit_ct['10%']:
            resultados[col] = 'Estacionário ao Redor de uma Tendência Linear'
            continue

        # Etapa 2 - Teste com constante apenas (regression='c')
        resultado_c = adfuller(serie, regression='c', autolag='AIC')
        stat_c = resultado_c[0]
        crit_c = resultado_c[4]

        if stat_c < crit_c['10%']:
            resultados[col] = 'Estacionário ao Redor de uma Constante'
            continue

        # Etapa 3 - Teste sem constante nem tendência (regression='n')
        resultado_n = adfuller(serie, regression='n', autolag='AIC')
        stat_n = resultado_n[0]
        crit_n = resultado_n[4]

        if stat_n < crit_n['10%']:
            resultados[col] = 'Estacionário com Média Zero'
        else:
            resultados[col] = 'Passeio Aleatório sem Drift'

    # Converter para DataFrame
    resultados_df = pd.DataFrame.from_dict(resultados, orient='index', columns=['Processo Gerador'])

    return resultados_df